crates/re_edit_ui/src/marker_shape.rs

```diff
@@ -0,0 +1,85 @@
+use egui::NumExt as _;
+
+use re_types::components::MarkerShape;
+use re_viewer_context::ViewerContext;
+
+pub(crate) fn edit_marker_shape_ui(
+    ctx: &ViewerContext<'_>,
+    ui: &mut egui::Ui,
+    edit_marker: &mut MarkerShape,
+) -> egui::Response {
+    let marker_text = edit_marker.to_string(); // TODO(emilk): Show marker shape in the selected text
+
+    let item_width = 100.0;
+
+    let outer_response = egui::ComboBox::from_id_source("marker_shape")
+        .selected_text(marker_text)
+        .width(
+            ui.available_width()
+                .at_most(item_width + ui.spacing().menu_margin.sum().x),
+        )
+        .height(320.0)
+        .show_ui(ui, |ui| {
+            // workaround to force `ui.max_rect()` to reflect the content size
+            ui.set_width(item_width);
+
+            let background_x_range = (ui.max_rect() + ui.spacing().menu_margin).x_range();
+
+            let list_ui = |ui: &mut egui::Ui| {
+                let mut combined_response: Option<egui::Response> = None;
+                for marker in MarkerShape::ALL {
+                    let mut response = ctx
+                        .re_ui
+                        .list_item()
+                        .selected(*edit_marker == marker)
+                        .show_flat(
+                            ui,
+                            re_ui::list_item::LabelContent::new(marker.to_string())
+                                .min_desired_width(item_width)
+                                .with_icon_fn(|_re_ui, ui, rect, visuals| {
+                                    paint_marker(ui, marker.into(), rect, visuals.text_color());
+                                }),
+                        );
+
+                    if response.clicked() {
+                        *edit_marker = marker;
+                        response.changed = true;
+                    }
+
+                    combined_response = Some(match combined_response {
+                        Some(combined_response) => combined_response.union(response),
+                        None => response,
+                    });
+                }
+                combined_response.expect("At least one marker shape should be available")
+            };
+
+            re_ui::full_span::full_span_scope(ui, background_x_range, |ui| {
+                re_ui::list_item::list_item_scope(ui, "marker_shape", list_ui)
+            })
+        });
+
+    outer_response.inner.unwrap_or(outer_response.response)
```

Comment:
That makes sense, but the thing that's really important here to me is the `changed` flag on the response since this causes a store write. If I pass the outer response, I won't get that flag.
Any suggestions how to deal with that? Is the design of relying on the `changed` response fundamentally flawed?

Reply:
Just call [`response.mark_changed()`](https://docs.rs/egui/latest/egui/struct.Response.html#method.mark_changed) on the outer response as needed

Reply:
I faced the same problem in a personal project. I ended up having a `changed: bool` returned value passed around with lots of `changed |= ui.thingy().changed();` in the implementation. Not really nice, but at least explicit.